services/api/service.go

```diff
@@ -386,6 +405,80 @@ func (api *RelayAPI) startValidatorRegistrationDBProcessor() {
 	}
 }
 
+// simulateBlock sends a request for a block simulation to blockSimRateLimiter.
+func (api *RelayAPI) simulateBlock(opts blockSimOptions) error {
+	t := time.Now()
+	simErr := api.blockSimRateLimiter.send(opts.ctx, opts.req, opts.highPrio)
+	log := opts.log.WithFields(logrus.Fields{
+		"duration":   time.Since(t).Seconds(),
+		"numWaiting": api.blockSimRateLimiter.currentCounter(),
+	})
+	if simErr != nil {
+		log.WithError(simErr).Error("block validation failed")
+		return simErr
+	}
+	log.Info("block validation successful")
+	return nil
+}
+
+// Modifies the status of a set of builders who have matching collateral IDs.
+// This function continues to try demoting even if an error is encountered for
+// public key.
+func (api *RelayAPI) setStatusByCollateralID(builderPubkey string, status common.BuilderStatus) {
+	// Fetch builder collateral_id.
+	builder, err := api.db.GetBlockBuilderByPubkey(builderPubkey)
+	if err != nil {
+		api.log.WithError(err).Error("unable to get builder from database")
+	}
+
+	// Fetch all builder pubkeys using the collateral_id.
+	builderPubkeys, err := api.db.GetBlockBuilderPubkeysByCollateralID(builder.CollateralID)
+	if err != nil {
+		api.log.WithError(err).Error("unable to get builders from collateral id")
+	}
+
+	// Demote all the pubkeys in both redis and the db.
+	for _, pubkey := range builderPubkeys {
+		err := api.datastore.SetBlockBuilderStatus(pubkey, status)
+		if err != nil {
+			api.log.WithError(err).Error("failed to set block builder status in the datastore")
+		}
+	}
+}
+
+// startOptimisticBlockProcessor keeps listening on the channel and validating incoming blocks asynchronously.
+func (api *RelayAPI) startOptimisticBlockProcessor() {
+	for opts := range api.optimisticBlockC {
+		builderPubkey := opts.req.Message.BuilderPubkey.String()
+		builderStatus, err := api.redis.GetBlockBuilderStatus(builderPubkey)
+		opts.log.WithFields(logrus.Fields{
+			"builderStatus": builderStatus.String(),
+		})
+		if err != nil {
+			opts.log.WithError(err).Error("could not get block builder status (defaulting to low-prio)")
+		}
+
+		// Skip simulation if the status is no longer OptimisticActive.
+		if builderStatus != common.OptimisticActive {
```

Comment:
I don't completely understand this. For example, can the status be OptimisticLocked and we want to demote if the block was invalid?

Reply:
oh you are right, i was thinking about it as if the status was OptimisticLocked, then the block is already being simulated in the getPayload handler, but as you point out, it could be a different block here in the startOptimisticBlockProcessor, so we still should process it. nice. what about the case where the status is `OptimisticDemoted`. there is not much use in continuing to simulate any blocks from them optimistically, right? 

changing it to if builderStatus == common.OptimisticDemoted { for now, but let me know if i am missing something!

Reply:
`if builderStatus == common.OptimisticDemoted` makes sense but now I'm tempted to remove those lines of code 😂  Partly to lean on the conservative side of things, and partly to simplify the logic. The additional cost of simulation in the instant a builder flips from optimisticActive to optimisticDemoted (which itself should rarely happen) is negligible.

Reply:
totally fair. removing in the name of simplicity!